File: packages/php/src/Mismatch.php
<?php

declare(strict_types=1);

namespace Xyp;

/**
 * One response field that did not fit the SDK's model. The field kept its
 * default value and the call still succeeded.
 *
 * The raw value is citizen data, so nothing prints it: not __toString, not
 * var_dump, not print_r. Ask for it with value().
 */
final class Mismatch implements \Stringable
{
    /**
     * @internal built by the decoder
     *
     * @param string $path    e.g. "listData[1].year"; "" is the response itself
     * @param string $problem what did not fit, e.g. "not a valid int"
     */
    public function __construct(
        public readonly string $path,
        public readonly string $problem,
        #[\SensitiveParameter]
        private readonly mixed $value,
    ) {}

    /** The raw value XYP sent for this field. */
    public function value(): mixed
    {
        return $this->value;
    }

    /** "path (problem)", never the value. */
    public function __toString(): string
    {
        return ($this->path === '' ? '<response>' : $this->path) . ' (' . $this->problem . ')';
    }

    /**
     * @return array{path: string, problem: string}
     */
    public function __debugInfo(): array
    {
        return ['path' => $this->path, 'problem' => $this->problem];
    }
}

File: packages/java/src/main/java/io/github/baljinnyamday/xyp/Mismatch.java
package io.github.baljinnyamday.xyp;

/**
 * One response field that did not fit the SDK's model. The field was decoded as {@code null} (an
 * empty list for a list) and the call still succeeded.
 *
 * <p>{@link #toString()} prints the path and the problem, never the value: the value is citizen
 * data and stays out of logs unless the caller asks for it through {@link #value()}.
 */
public final class Mismatch {

  private final String path;
  private final String problem;
  private final Object value;

  Mismatch(String path, String problem, Object value) {
    this.path = path;
    this.problem = problem;
    this.value = value;
  }

  /**
   * Where the field is in the response.
   *
   * @return e.g. {@code "listData[1].year"}; {@code ""} is the response itself
   */
  public String path() {
    return path;
  }

  /**
   * What did not fit.
   *
   * @return e.g. {@code "not a valid int"} or {@code "expected an object"}
   */
  public String problem() {
    return problem;
  }

  /**
   * The raw value XYP sent for this field: a {@code String}, or an unmodifiable {@code Map} or
   * {@code List} of the response tree.
   *
   * @return the raw value
   */
  public Object value() {
    return value;
  }

  /**
   * The path and the problem, never the value.
   *
   * @return e.g. {@code "age (not a valid int)"}, or {@code "<response> (expected an object)"}
   */
  @Override
  public String toString() {
    return (path.isEmpty() ? "<response>" : path) + " (" + problem + ")";
  }
}
